baba.cc

```diff
@@ -302,50 +302,60 @@ class Runner {
 
             if (info.ssi_signo == SIGCHLD) {
                 // One or more children (not descendants!) finished.
-                pid_t pid;
+                
                 int stat;
-                while ((pid = waitpid(0, &stat, WNOHANG)) >= 0) {
-                    bool exited = WIFEXITED(stat);
-                    int status = WEXITSTATUS(stat);
-                    bool is_first_child = pid == first_child_pid_;
-                    bool tracked = (track_fails_ == GROUP ||
-                        (track_fails_ == CHILD && is_first_child));
-                    if (!exited || status) {
-                        if (is_first_child)
-                            first_child_status_ = status;
-                        if (tracked)
-                            failed_count_++;
-                        if (log_level_ >= VERBOSE) {
-                            if (exited) {
-                                std::cerr
-                                    << pid
-                                    << " exited with status "
-                                    << status
-                                    << std::endl;
-                            }
-                            else {
-                                std::cerr
-                                    << pid
-                                    << " terminated early"
-                                    << std::endl;
-                            }
-                        }
-                    } else if (log_level_ == TRACE)
-                        std::cerr << pid <<
-                            " successfully exited" <<
-                            std::endl;
+                pid_t pid = waitpid(0, &stat, WNOHANG);
+                while (pid > 0) {
+                    _handle_stat(pid, stat);
+                    pid = waitpid(0, &stat, WNOHANG);
```

Comment:
That's what the earlier version did, and getting back to it I was shortly confused by the assignment in a conditional.
Most likely because I've been using other languages lately: it's obviously idiomatic.